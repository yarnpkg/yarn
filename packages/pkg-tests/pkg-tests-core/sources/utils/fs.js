/* @flow */

import type {Gzip} from 'zlib';

const fs = require('fs-extra');
const klaw = require('klaw');
const path = require('path');
const tarFs = require('tar-fs');
const tmp = require('tmp');
const zlib = require('zlib');

const miscUtils = require('./misc');

exports.walk = function walk(
  source: string,
  {filter, relative = false}: {|filter?: Array<string>, relative?: boolean|} = {},
): Promise<Array<string>> {
  return new Promise((resolve, reject) => {
    const paths = [];

    const walker = klaw(source, {
      filter: itemPath => {
        if (!filter) {
          return true;
        }

        const stat = fs.statSync(itemPath);

        if (stat.isDirectory()) {
          return true;
        }

        const relativePath = path.relative(source, itemPath);

        if (miscUtils.filePatternMatch(relativePath, filter)) {
          return true;
        }

        return false;
      },
    });

    walker.on('data', ({path: itemPath}) => {
      const relativePath = path.relative(source, itemPath);

      if (!filter || miscUtils.filePatternMatch(relativePath, filter)) {
        paths.push(relative ? relativePath : itemPath);
      }

      // This item has been accepted only because it's a directory; it doesn't match the filter
      return;
    });

    walker.on('end', () => {
      resolve(paths);
    });
  });
};

exports.packToStream = function packToStream(
  source: string,
  // $FlowFixMe Flow doesn't support null default parameters in exact types
  {virtualPath = null}: {|virtualPath?: ?string|} = {},
): Gzip {
  if (virtualPath) {
    if (!path.isAbsolute(virtualPath)) {
      throw new Error('The virtual path has to be an absolute path');
    } else {
      virtualPath = path.resolve(virtualPath);
    }
  }

  const zipperStream = zlib.createGzip();

  const packStream = tarFs.pack(source, {
    map: header => {
      if (true) {
        header.name = path.resolve('/', header.name);
        header.name = path.relative('/', header.name);
      }

      if (virtualPath) {
        header.name = path.resolve('/', virtualPath, header.name);
        header.name = path.relative('/', header.name);
      }

      return header;
    },
  });

  packStream.pipe(zipperStream);

  packStream.on('error', error => {
    zipperStream.emit('error', error);
  });

  return zipperStream;
};

exports.packToFile = function packToFile(target: string, source: string, options: *): Promise<void> {
  const tarballStream = fs.createWriteStream(target);

  const packStream = exports.packToStream(source, options);
  packStream.pipe(tarballStream);

  return new Promise((resolve, reject) => {
    tarballStream.on('error', error => {
      reject(error);
    });

    packStream.on('error', error => {
      reject(error);
    });

    tarballStream.on('close', () => {
      resolve();
    });
  });
};

exports.createTemporaryFolder = function createTemporaryFolder(): Promise<string> {
  return new Promise((resolve, reject) => {
    tmp.dir({unsafeCleanup: true}, (error, dirPath) => {
      if (error) {
        reject(error);
      } else {
        resolve(dirPath);
      }
    });
  });
};

exports.createTemporaryFile = async function createTemporaryFile(filePath: string): Promise<string> {
  if (filePath) {
    if (path.normalize(filePath).match(/^(\.\.)?\//)) {
      throw new Error('A temporary file path must be a forward path');
    }

    const folderPath = await exports.createTemporaryFolder();
    return path.resolve(folderPath, filePath);
  } else {
    return new Promise((resolve, reject) => {
      tmp.file({discardDescriptor: true}, (error, filePath) => {
        if (error) {
          reject(error);
        } else {
          resolve(filePath);
        }
      });
    });
  }
};

exports.writeFile = async function writeFile(target: string, body: string | Buffer): Promise<void> {
  await fs.mkdirp(path.dirname(target));
  await fs.writeFile(target, body);
};

exports.readFile = function readFile(source: string, encoding: ?string = null): Promise<any> {
  return fs.readFile(source, encoding);
};

exports.writeJson = function writeJson(target: string, object: any): Promise<void> {
  return exports.writeFile(target, JSON.stringify(object));
};

exports.readJson = async function readJson(source: string): Promise<any> {
  const fileContent = await exports.readFile(source, 'utf8');

  try {
    return JSON.parse(fileContent);
  } catch (error) {
    throw new Error(`Invalid json file (${source})`);
  }
};
