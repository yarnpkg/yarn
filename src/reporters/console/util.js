/* @flow */

import type {Stdout} from '../types.js';

const readline = require('readline');

export function clearLine(stdout: Stdout) {
  readline.clearLine(stdout, 0);
  readline.cursorTo(stdout, 0);
}

export function toStartOfLine(stdout: Stdout) {
  readline.cursorTo(stdout, 0);
}

export function writeOnNthLine(stdout: Stdout, n: number, msg: string) {
  if (n == 0) {
    clearLine(stdout);
    stdout.write(msg);
    return;
  }
  readline.cursorTo(stdout, 0);
  readline.moveCursor(stdout, 0, -n);
  readline.clearLine(stdout, 0);
  stdout.write(msg);
  readline.cursorTo(stdout, 0);
  readline.moveCursor(stdout, 0, n);
}

export function clearNthLine(stdout: Stdout, n: number) {
  if (n == 0) {
    clearLine(stdout);
    return;
  }
  readline.cursorTo(stdout, 0);
  readline.moveCursor(stdout, 0, -n);
  readline.clearLine(stdout, 0);
  readline.moveCursor(stdout, 0, n);
}
