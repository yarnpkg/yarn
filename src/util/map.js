/* @flow */

type Return<T> = T | Object;

export default function nullify<T>(obj?: Return<T> = {}): Return<T> {
  if (Array.isArray(obj)) {
    for (const item of obj) {
      nullify(item);
    }
  } else if (obj !== null && typeof obj === 'object' || typeof obj === 'function') {
    Object.setPrototypeOf(obj, null);

    // for..in can only be applied to 'object', not 'function'
    if (typeof obj === 'object') {
      for (const key in obj) {
        nullify(obj[key]);
      }
    }
  }

  return obj;
}
