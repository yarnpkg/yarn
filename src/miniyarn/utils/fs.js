import * as Fse from 'fs-extra';
import * as Fs from 'fs';
import gunzipMaybe from 'gunzip-maybe';
import klaw from 'klaw';
import Path from 'path';
import rcopy from 'recursive-copy';
import tarFs from 'tar-fs';
import tar from 'tar-stream';
import tmp from 'tmp';
import {createGzip} from 'zlib';

import * as miscUtils from 'miniyarn/utils/misc';
import * as pathUtils from 'miniyarn/utils/path';
import * as parseUtils from 'miniyarn/utils/parse';

let isWindows = null;

export async function detectWindows() {
  if (isWindows !== null) return isWindows;

  if (await exists(`c:/`)) return (isWindows = true);

  if ((await exists(`/proc/version`)) && (await readFile(`/proc/version`, `utf8`)).match(/Microsoft|WSL/))
    return (isWindows = true);

  return (isWindows = false);
}

export async function cwd(path, callback) {
  let cwd = process.cwd();
  process.chdir(path);

  try {
    return await callback();
  } finally {
    process.chdir(cwd);
  }
}

export async function symlink(source, destination) {
  return await new Promise((resolve, reject) => {
    Fs.symlink(source, destination, error => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}

export async function ensureSymlink(source, destination) {
  return await new Promise((resolve, reject) => {
    Fse.ensureSymlink(source, destination, error => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}

export async function isDirectory(path) {
  return await new Promise((resolve, reject) => {
    Fs.lstat(path, (error, stat) => {
      if (error) {
        reject(error);
      } else {
        resolve(stat.isDirectory());
      }
    });
  });
}

export async function createTemporaryFolder() {
  return new Promise((resolve, reject) => {
    tmp.dir({unsafeCleanup: true}, (error, path) => {
      if (error) {
        reject(error);
      } else {
        resolve(path);
      }
    });
  });
}

export async function createTemporaryFile(filePath) {
  if (filePath) {
    if (!pathUtils.isForward(filePath)) throw new Error(`A temporary file path must be a forward path`);

    let folderPath = await createTemporaryFolder();

    return pathUtils.resolve(folderPath, filePath);
  } else {
    return await new Promise((resolve, reject) => {
      tmp.file((error, path) => {
        if (error) {
          reject(error);
        } else {
          resolve(path);
        }
      });
    });
  }
}

export async function walk(path, {filter} = {}) {
  return await new Promise((resolve, reject) => {
    let paths = [];

    let walker = klaw(path, {
      filter: itemPath => {
        if (!filter) return true;

        if (Fse.statSync(path).isDirectory()) return true;
        s;
        let relativePath = pathUtils.relative(path, itemPath);

        if (miscUtils.filePatternMatch(relativePath, filter)) return true;

        return false;
      },
    });

    walker.on(`data`, ({path: itemPath}) => {
      if (!filter) return void paths.push(itemPath);

      let relativePath = pathUtils.relative(path, itemPath);

      if (miscUtils.filePatternMatch(relativePath, filter)) return void paths.push(itemPath);

      return; // This item has been accepted only because it's a directory; it doesn't match the filter
    });

    walker.on(`end`, () => {
      resolve(paths);
    });
  });
}

export async function mkdirp(path) {
  return await new Promise((resolve, reject) => {
    Fse.ensureDir(path, error => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}

export async function rm(path) {
  if (path === ``) throw new Error(`Cannot rm an empty path`);

  if (path === `/`) throw new Error(`Cannot rm your whole / directory - something's very wrong`);

  if (!await exists(path)) return;

  return await new Promise((resolve, reject) => {
    Fse.remove(path, err => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

export async function getMTime(path) {
  return await new Promise((resolve, reject) => {
    Fs.lstat(path, (error, stat) => {
      if (error) {
        reject(error);
      } else {
        resolve(stat.mtime);
      }
    });
  });
}

export async function createFileWriter(path) {
  await mkdirp(pathUtils.dirname(path));

  let stream = Fs.createWriteStream(path);

  stream.promise = new Promise((resolve, reject) => {
    stream.on(`finish`, () => {
      resolve();
    });

    stream.on(`error`, error => {
      reject(error);
    });
  });

  return stream;
}

export function createFileReader(path) {
  let stream = Fs.createReadStream(path);

  stream.promise = new Promise((resolve, reject) => {
    stream.on(`error`, error => {
      reject(error);
    });

    stream.on(`end`, () => {
      resolve();
    });
  });

  return stream;
}

export async function exists(path) {
  return await new Promise((resolve, reject) => {
    Fs.exists(path, status => {
      resolve(status);
    });
  });
}

export async function cp(source, destination, {filter = []} = {}) {
  await mkdirp(pathUtils.dirname(destination));
  await rcopy(source, destination, {dot: true, filter: path => miscUtils.filePatternMatch(path, filter)});
}

export async function mv(source, target) {
  // Windows defender often lock files after they are created, which makes it impossible to move them.
  // We could have a mechanism of "try again", but it seems simpler to just use a copy operation instead.
  // Note that using cp instead of mv is not strictly comparable since, unfortunately, it's not atomic. Nothing we can do, tho :(

  if (await detectWindows()) {
    await cp(source, target);
    await rm(source);
  } else {
    await mkdirp(pathUtils.dirname(target));

    return await new Promise((resolve, reject) => {
      Fs.rename(source, target, err => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      });
    });
  }
}

export async function extract(target, tarballPath, {stripLeadingFolder = false} = {}) {
  return await new Promise((resolve, reject) => {
    function map(header) {
      if (header.name[0] === `/`) header.name = header.name.replace(/^\/+/, ``);

      if (stripLeadingFolder) header.name = header.name.replace(/^[^\/]+\//, ``);

      return header;
    }

    let unzipper = gunzipMaybe();
    Fs.createReadStream(tarballPath).pipe(unzipper);

    let extractor = tarFs.extract(target, {map});
    unzipper.pipe(extractor);

    extractor.on(`error`, error => {
      reject(error);
    });

    extractor.on(`finish`, () => {
      resolve();
    });
  });
}

export function packToStream(path, {virtualPath = null} = {}) {
  if (virtualPath) {
    if (!pathUtils.isAbsolute(virtualPath)) {
      throw new Error(`The virtual path has to be an absolute path`);
    } else {
      virtualPath = pathUtils.resolve(virtualPath);
    }
  }

  let zipperStream = createGzip();

  let packStream = tarFs.pack(path, {
    map: header => {
      if (true) {
        header.name = pathUtils.resolve(`/`, header.name);
        header.name = pathUtils.relative(`/`, header.name);
      }

      if (virtualPath) {
        header.name = pathUtils.resolve(`/`, virtualPath, header.name);
        header.name = pathUtils.relative(`/`, header.name);
      }

      return header;
    },
  });

  packStream.pipe(zipperStream);

  packStream.on(`error`, error => {
    zipperStream.emit(`error`, error);
  });

  return zipperStream;
}

export async function packToFile(target, directoryPath, {...options} = {}) {
  let tarballStream = Fs.createWriteStream(target);

  let packStream = packToStream(directoryPath, options);
  packStream.pipe(tarballStream);

  return await new Promise((resolve, reject) => {
    tarballStream.on(`error`, error => {
      reject(error);
    });

    packStream.on(`error`, error => {
      reject(error);
    });

    tarballStream.on(`close`, () => {
      resolve();
    });
  });
}

export async function writeFile(path, body) {
  await mkdirp(pathUtils.dirname(path));

  return await new Promise((resolve, reject) => {
    Fs.writeFile(path, body, err => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

export async function readFile(path, encoding = null) {
  return await new Promise((resolve, reject) => {
    Fs.readFile(path, encoding, (err, body) => {
      if (err) {
        reject(err);
      } else {
        resolve(body);
      }
    });
  });
}

export async function readJson(path) {
  let fileContent = await readFile(path, `utf8`);

  return parseUtils.parseJson(fileContent);
}

export async function readDirectory(path) {
  return await new Promise((resolve, reject) => {
    Fs.readdir(path, (err, entries) => {
      if (err) {
        reject(err);
      } else {
        resolve(entries);
      }
    });
  });
}

export class Handler {
  constructor(path, {temporary = false} = {}) {
    this.path = path;

    this.temporary = temporary;
  }

  get() {
    return this.path;
  }

  async steal() {
    if (this.temporary) return this.path;

    let destination = (await isDirectory(this.path)) ? await createTemporaryFolder() : await createTemporaryFile();

    await cp(this.path, destination);

    return destination;
  }
}
