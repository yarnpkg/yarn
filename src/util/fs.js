/* @flow */

import BlockingQueue from './blocking-queue.js';
import * as promise from './promise.js';
import {promisify} from './promise.js';
import map from './map.js';

const path = require('path');
const fs = require('fs');

export const lockQueue = new BlockingQueue('fs lock');

export const readFileBuffer = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);
export const readlink = promisify(fs.readlink);
export const realpath = promisify(fs.realpath);
export const readdir = promisify(fs.readdir);
export const rename = promisify(fs.rename);
export const access = promisify(fs.access);
export const stat = promisify(fs.stat);
export const unlink = promisify(require('rimraf'));
export const mkdirp = promisify(require('mkdirp'));
export const exists = promisify(fs.exists, true);
export const lstat = promisify(fs.lstat);
export const chmod = promisify(fs.chmod);

const fsSymlink = promisify(fs.symlink);
const invariant = require('invariant');
const stripBOM = require('strip-bom');

const noop = () => {};

export type CopyQueueItem = {
  src: string,
  dest: string,
  onFresh?: ?() => void,
  onDone?: ?() => void,
};

type CopyQueue = Array<CopyQueueItem>;

type CopyFileAction = {
  type: 'file',
  src: string,
  dest: string,
  atime: number,
  mtime: number,
  mode: number
};

type CopySymlinkAction = {
  type: 'symlink',
  dest: string,
  linkname: string,
};

type CopyActions = Array<CopyFileAction | CopySymlinkAction>;

type PossibleExtraneous = void | false | Iterable<string>;

type CopyOptions = {
  onProgress: (dest: string) => void,
  onStart: (num: number) => void,
  possibleExtraneous: PossibleExtraneous,
  ignoreBasenames: Array<string>,
};

async function buildActionsForCopy(
  queue: CopyQueue,
  events: CopyOptions,
  possibleExtraneousSeed: PossibleExtraneous,
): Promise<CopyActions> {
  const possibleExtraneous: Set<string> = new Set(possibleExtraneousSeed || []);
  const noExtraneous = possibleExtraneousSeed === false;
  const files: Set<string> = new Set();

  // initialise events
  for (const item of queue) {
    const onDone = item.onDone;
    item.onDone = () => {
      events.onProgress(item.dest);
      if (onDone) {
        onDone();
      }
    };
  }
  events.onStart(queue.length);

  // start building actions
  const actions: CopyActions = [];

  // custom concurrency logic as we're always executing stacks of 4 queue items
  // at a time due to the requirement to push items onto the queue
  while (queue.length) {
    const items = queue.splice(0, 4);
    await Promise.all(items.map(build));
  }

  // remove all extraneous files that weren't in the tree
  if (!noExtraneous) {
    for (const loc of possibleExtraneous) {
      if (!files.has(loc)) {
        await unlink(loc);
      }
    }
  }

  return actions;

  //
  async function build(data): Promise<void> {
    const {src, dest} = data;
    const onFresh = data.onFresh || noop;
    const onDone = data.onDone || noop;
    files.add(dest);

    if (events.ignoreBasenames.indexOf(path.basename(src)) >= 0) {
      // ignored file
      return;
    }

    const srcStat = await lstat(src);
    let srcFiles;

    if (srcStat.isDirectory()) {
      srcFiles = await readdir(src);
    }

    if (await exists(dest)) {
      const destStat = await lstat(dest);

      const bothSymlinks = srcStat.isSymbolicLink() && destStat.isSymbolicLink();
      const bothFolders = srcStat.isDirectory() && destStat.isDirectory();
      const bothFiles = srcStat.isFile() && destStat.isFile();

      if (srcStat.mode !== destStat.mode) {
        try {
          await access(dest, srcStat.mode);
        } catch (err) {
          // EINVAL access errors sometimes happen which shouldn't because node shouldn't be giving
          // us modes that aren't valid. investigate this, it's generally safe to proceed.
        }
      }

      if (bothFiles && srcStat.size === destStat.size && +srcStat.mtime === +destStat.mtime) {
        // we can safely assume this is the same file
        onDone();
        return;
      }

      if (bothSymlinks && await readlink(src) === await readlink(dest)) {
        // if both symlinks are the same then we can continue on
        onDone();
        return;
      }

      if (bothFolders && !noExtraneous) {
        // mark files that aren't in this folder as possibly extraneous
        const destFiles = await readdir(dest);
        invariant(srcFiles, 'src files not initialised');

        for (const file of destFiles) {
          if (srcFiles.indexOf(file) < 0) {
            const loc = path.join(dest, file);
            possibleExtraneous.add(loc);

            if ((await lstat(loc)).isDirectory()) {
              for (const file of await readdir(loc)) {
                possibleExtraneous.add(path.join(loc, file));
              }
            }
          }
        }
      }
    }

    if (srcStat.isSymbolicLink()) {
      onFresh();
      const linkname = await readlink(src);
      actions.push({
        type: 'symlink',
        dest,
        linkname,
      });
      onDone();
    } else if (srcStat.isDirectory()) {
      await mkdirp(dest);

      const destParts = dest.split(path.sep);
      while (destParts.length) {
        files.add(destParts.join(path.sep));
        destParts.pop();
      }

      // push all files to queue
      invariant(srcFiles, 'src files not initialised');
      let remaining = srcFiles.length;
      if (!remaining) {
        onDone();
      }
      for (const file of srcFiles) {
        queue.push({
          onFresh,
          src: path.join(src, file),
          dest: path.join(dest, file),
          onDone: () => {
            if (--remaining === 0) {
              onDone();
            }
          },
        });
      }
    } else if (srcStat.isFile()) {
      onFresh();
      actions.push({
        type: 'file',
        src,
        dest,
        atime: srcStat.atime,
        mtime: srcStat.mtime,
        mode: srcStat.mode,
      });
      onDone();
    } else {
      throw new Error(`unsure how to copy this: ${src}`);
    }
  }
}

export function copy(src: string, dest: string): Promise<void> {
  return copyBulk([{src, dest}]);
}

export async function copyBulk(
  queue: CopyQueue,
  _events?: {
    onProgress?: ?(dest: string) => void,
    onStart?: ?(num: number) => void,
    possibleExtraneous?: PossibleExtraneous,
    ignoreBasenames?: Array<string>,
  },
): Promise<void> {
  const events: CopyOptions = {
    onStart: (_events && _events.onStart) || noop,
    onProgress: (_events && _events.onProgress) || noop,
    possibleExtraneous: _events ? _events.possibleExtraneous : [],
    ignoreBasenames: (_events && _events.ignoreBasenames) || [],
  };

  const actions: CopyActions = await buildActionsForCopy(queue, events, events.possibleExtraneous);
  events.onStart(actions.length);

  const fileActions = actions.filter((action): boolean => action.type === 'file');
  await promise.queue(fileActions, (data): Promise<void> => new Promise((resolve, reject) => {
    const readStream = fs.createReadStream(data.src);
    const writeStream = fs.createWriteStream(data.dest, {mode: data.mode});

    readStream.on('error', reject);
    writeStream.on('error', reject);

    writeStream.on('open', function() {
      readStream.pipe(writeStream);
    });

    writeStream.once('finish', function() {
      fs.utimes(data.dest, data.atime, data.mtime, function(err) {
        if (err) {
          reject(err);
        } else {
          events.onProgress(data.dest);
          resolve();
        }
      });
    });
  }), 4);

  // we need to copy symlinks last as the could reference files we were copying
  const symlinkActions = actions.filter((action): boolean => action.type === 'symlink');
  await promise.queue(symlinkActions, (data): Promise<void> => {
    return symlink(path.resolve(path.dirname(data.dest), data.linkname), data.dest);
  });
}

function _readFile(loc: string, encoding: string): Promise<any> {
  return new Promise((resolve, reject) => {
    fs.readFile(loc, encoding, function(err, content) {
      if (err) {
        reject(err);
      } else {
        resolve(content);
      }
    });
  });
}

export function readFile(loc: string): Promise<string> {
  return _readFile(loc, 'utf8').then(normalizeOS);
}

export function readFileRaw(loc: string): Promise<Buffer> {
  return _readFile(loc, 'binary');
}

export async function readFileAny(files: Array<string>): Promise<?string> {
  for (const file of files) {
    if (await exists(file)) {
      return readFile(file);
    }
  }
  return null;
}

export async function readJson(loc: string): Promise<Object> {
  const file = await readFile(loc);
  try {
    return map(JSON.parse(stripBOM(file)));
  } catch (err) {
    err.message = `${loc}: ${err.message}`;
    throw err;
  }
}

export async function find(filename: string, dir: string): Promise<string | false> {
  const parts = dir.split(path.sep);

  while (parts.length) {
    const loc = parts.concat(filename).join(path.sep);

    if (await exists(loc)) {
      return loc;
    } else {
      parts.pop();
    }
  }

  return false;
}

export async function symlink(src: string, dest: string): Promise<void> {
  try {
    const stats = await lstat(dest);

    if (stats.isSymbolicLink() && await exists(dest)) {
      const resolved = await realpath(dest);
      if (resolved === src) {
        return;
      }
    }

    await unlink(dest);
  } catch (err) {
    if (err.code !== 'ENOENT') {
      throw err;
    }
  }

  try {
    if (process.platform === 'win32') {
      // use directory junctions if possible on win32, this requires absolute paths
      await fsSymlink(src, dest, 'junction');
    } else {
      // use relative paths otherwise which will be retained if the directory is moved
      const relative = path.relative(path.dirname(dest), src);
      await fsSymlink(relative, dest);
    }
  } catch (err) {
    if (err.code === 'EEXIST') {
      // race condition
      await symlink(src, dest);
    } else {
      throw err;
    }
  }
}

export type WalkFiles = Array<{
  relative: string,
  absolute: string,
  basename: string,
  mtime: number,
}>;

export async function walk(
  dir: string,
  relativeDir?: ?string,
  ignoreBasenames?: Set<string> = new Set(),
): Promise<WalkFiles> {
  let files = [];

  let filenames = await readdir(dir);
  if (ignoreBasenames.size) {
    filenames = filenames.filter((name) => !ignoreBasenames.has(name));
  }

  for (const name of filenames) {
    const relative = relativeDir ? path.join(relativeDir, name) : name;
    const loc = path.join(dir, name);
    const stat = await lstat(loc);

    files.push({
      relative,
      basename: name,
      absolute: loc,
      mtime: +stat.mtime,
    });

    if (stat.isDirectory()) {
      files = files.concat(await walk(loc, relative, ignoreBasenames));
    }
  }

  return files;
}

export function normalizeOS(body: string): string {
  return body.replace(/\r\n/g, '\n');
}
