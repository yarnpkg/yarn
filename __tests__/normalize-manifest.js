/* @flow */
/* eslint max-len: 0 */

import normalizeManifest from '../src/util/normalize-manifest/index.js';
import NoopReporter from '../src/reporters/base-reporter.js';
import Config from '../src/config.js';
import map from '../src/util/map.js';
import * as util from '../src/util/normalize-manifest/util.js';
import * as fs from '../src/util/fs.js';

const nativeFs = require('fs');
const path     = require('path');

jasmine.DEFAULT_TIMEOUT_INTERVAL = 60000;

const fixturesLoc = path.join(__dirname, 'fixtures', 'normalize-manifest');

for (const name of nativeFs.readdirSync(fixturesLoc)) {
  if (name[0] === '.') {
    continue;
  }

  const loc = path.join(fixturesLoc, name);

  test(name, async () => {
    const actualWarnings   = [];
    const expectedWarnings = await fs.readJson(path.join(loc, 'warnings.json'));

    const reporter = new NoopReporter();

    // $FlowFixMe: investigate
    reporter.warn = function(msg) {
      actualWarnings.push(msg);
    };

    const config = new Config(reporter);
    await config.init({cwd: loc});

    let actual   = await fs.readJson(path.join(loc, 'actual.json'));
    const expected = await fs.readJson(path.join(loc, 'expected.json'));

    let isRoot = actual._root;
    if (isRoot == null) {
      isRoot = true;
    } else {
      delete actual._root;
    }

    const error = expected._error;
    if (error) {
      delete expected._error;
    }

    try {
      actual = await normalizeManifest(actual, loc, config, isRoot);
    } catch (err) {
      if (error && err.message === error) {
        return;
      } else {
        throw err;
      }
    }

    if (error) {
      throw new Error(`Expected to throw error: ${error}`);
    }

    expect(map(actual)).toEqual(expand(expected));
    expect(actualWarnings).toEqual(expectedWarnings);
  });
}

test('util.stringifyPerson', () => {
  expect(util.stringifyPerson({name: 'Sebastian McKenzie'})).toEqual('Sebastian McKenzie');
  expect(util.stringifyPerson({name: 'Sebastian McKenzie', email: 'sebmck@gmail.com'})).toEqual('Sebastian McKenzie <sebmck@gmail.com>');
  expect(util.stringifyPerson({email: 'sebmck@gmail.com'})).toEqual('<sebmck@gmail.com>');
  expect(util.stringifyPerson({name: 'Sebastian McKenzie', email: 'sebmck@gmail.com', url: 'https://sebmck.com'})).toEqual('Sebastian McKenzie <sebmck@gmail.com> (https://sebmck.com)');
});

test('util.parsePerson', () => {
  expect(util.parsePerson({}), {});
  expect(util.parsePerson('Sebastian McKenzie')).toEqual({name: 'Sebastian McKenzie'});
  expect(util.parsePerson(' <sebmck@gmail.com>')).toEqual({email: 'sebmck@gmail.com'});
  expect(util.parsePerson('Sebastian McKenzie <sebmck@gmail.com>')).toEqual({name: 'Sebastian McKenzie', email: 'sebmck@gmail.com'});
  expect(util.parsePerson('Sebastian McKenzie <sebmck@gmail.com> (https://sebmck.com)')).toEqual({name: 'Sebastian McKenzie', email: 'sebmck@gmail.com', url: 'https://sebmck.com'});
});

test('util.extractDescription', () => {
  expect(util.extractDescription('# header\n\ndescription here')).toEqual('description here');
  expect(util.extractDescription('# header\ndescription here')).toEqual('description here');
  expect(util.extractDescription('# header\ndescription here\nfoobar')).toEqual('description here foobar');
  expect(util.extractDescription('# header\ndescription here\n\nfoobar')).toEqual('description here');
  expect(util.extractDescription('')).toEqual(undefined);
  expect(util.extractDescription(null)).toEqual(undefined);
  expect(util.extractDescription(undefined)).toEqual(undefined);
});

// fill out expected and normalize paths
function expand<T>(expected: T): T {
  if (expected.man && Array.isArray(expected.man)) {
    expected = {...expected, man: normalizePaths(expected.man)};
  }

  if (expected.bin && typeof expected.bin === 'object') {
    expected = {...expected, bin: normalizePathDict(expected.bin)};
  }

  return expected;
}

function normalize(input: string): string {
  return path.normalize(input);
}

function normalizePath<T>(path: T): ?string {
  if (typeof path === 'string') {
    return normalize(path);
  } else {
    return null;
  }
}

function normalizePaths(paths: mixed): ?string[] {
  if (Array.isArray(paths)) {
    return paths.map((p) => {
      if (typeof p !== 'string') {
        throw new Error(`Expected string in paths, got ${JSON.stringify(paths)}`);
      }
      return normalize(p);
    });
  } else {
    return null;
  }
}

function normalizePathDict(paths: mixed): ?{ [key: string]: mixed } {
  const out = {};

  if (!paths || typeof paths !== 'object') {
    return null;
  }

  for (const prop in paths) {
    if (typeof paths[prop] === 'string') {
      out[prop] = normalizePath(paths[prop]);
    } else {
      out[prop] = paths[prop];
    }
  }

  return out;
}
