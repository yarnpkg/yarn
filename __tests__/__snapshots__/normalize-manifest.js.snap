// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`author object normalize 1`] = `
Array [
  "No license field",
]
`;

exports[`author string 1`] = `
Array [
  "No license field",
]
`;

exports[`authors file to contributors 1`] = `
Array [
  "No license field",
]
`;

exports[`bin string 1`] = `
Array [
  "foo: No license field",
]
`;

exports[`binding.gyp scripts.install 1`] = `
Array [
  "No license field",
]
`;

exports[`bugs string 1`] = `
Array [
  "No license field",
]
`;

exports[`bundledDependencies alias 1`] = `
Array [
  "No license field",
]
`;

exports[`clean version 1`] = `
Array [
  "No license field",
]
`;

exports[`clean version with v 1`] = `
Array [
  "No license field",
]
`;

exports[`contributors array expand 1`] = `
Array [
  "No license field",
]
`;

exports[`dedupe all trivial dependencies 1`] = `
Array [
  "No license field",
]
`;

exports[`dedupe dependencies 1`] = `Array []`;

exports[`dedupe dependencies root 1`] = `
Array [
  "No license field",
  "\\"dependencies\\" has dependency \\"foobar\\" with range \\"2.0.0\\" that collides with a dependency in \\"devDependencies\\" of the same name with version \\"1.0.0\\"",
]
`;

exports[`dedupe optionalDependencies 1`] = `Array []`;

exports[`dedupe optionalDependencies root 1`] = `
Array [
  "No license field",
  "\\"optionalDependencies\\" has dependency \\"foobar\\" with range \\"1.0.0\\" that collides with a dependency in \\"dependencies\\" of the same name with version \\"2.0.0\\"",
]
`;

exports[`dedupe single trivial dependency 1`] = `
Array [
  "No license field",
]
`;

exports[`engines array 1`] = `
Array [
  "No license field",
]
`;

exports[`explode directories.bin 1`] = `
Array [
  "No license field",
]
`;

exports[`explode directories.man 1`] = `
Array [
  "No license field",
]
`;

exports[`extract description 1`] = `
Array [
  "No license field",
]
`;

exports[`homepage string 1`] = `
Array [
  "No license field",
]
`;

exports[`hosted git repository url 1`] = `
Array [
  "No license field",
]
`;

exports[`keywords string 1`] = `
Array [
  "No license field",
]
`;

exports[`license warnings should not be thrown with global commands 1`] = `Array []`;

exports[`maintainers array expand 1`] = `
Array [
  "No license field",
]
`;

exports[`man string 1`] = `
Array [
  "No license field",
]
`;

exports[`name 1`] = `
Array [
  "foobar: No license field",
]
`;

exports[`readme 1`] = `
Array [
  "No license field",
]
`;

exports[`repository string 1`] = `
Array [
  "No license field",
]
`;

exports[`scoped bin string 1`] = `
Array [
  "@scope/foo: No license field",
]
`;

exports[`server scripts.start 1`] = `
Array [
  "No license field",
]
`;

exports[`valid license 1`] = `Array []`;

exports[`version 1`] = `
Array [
  "No license field",
]
`;

exports[`warn builtin module collision 1`] = `
Array [
  "fs: \\"fs\\" is also the name of a node core module",
  "fs: No license field",
]
`;

exports[`warn empty string invalid license 1`] = `
Array [
  "License should be a valid SPDX license expression",
]
`;

exports[`warn invalid license 1`] = `
Array [
  "License should be a valid SPDX license expression",
]
`;

exports[`warn typo 1`] = `
Array [
  "Potential typo \\"dependancies\\", did you mean \\"dependencies\\"?",
  "No license field",
]
`;
